File: Data Structures & Algorithms/merge-k-sorted-linked-lists/submission-0.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        // int numNonEmpty = lists.size();
        ListNode res(0);
        ListNode *curr = &res;

        while (1) {
            int minVal = 1000000;
            int minIndex = -1;
            ListNode* minList = nullptr;

            for (int i = 0; i < lists.size(); i++) {
                if (lists[i] && (lists[i]->val < minVal)) {
                    minVal = lists[i]->val;
                    minList = lists[i];
                    minIndex = i;
                }
            }

            if (minIndex >= 0) {
                lists[minIndex] = lists[minIndex]->next;
                minList->next = nullptr;
                curr->next = minList;
                curr = curr->next;
            }
            else {
                break;
            }
        }

        // if (curr) {
        //     curr->next = nullptr;
        // }

        return res.next;
    }
};

File: Data Structures & Algorithms/reverse-a-linked-list/submission-0.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        prev = None
        curr = head
        newHead = None

        while curr != None:
            # end of list, new head
            if curr.next == None:
                curr.next = prev
                newHead = curr
                break
            
            temp = curr.next
            curr.next = prev

            prev = curr
            curr = temp

        return newHead



        